# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/PeerConnectionImpl.cpp
# Commit: 7888e4be3d7e
# Full Hash: 7888e4be3d7e633739cf590e73b0a3af0ef3f9fc
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 5
# Description:
#   Bug 1654112 - Fix remote-inbound-rtp.packetsReceived. r=bwc
#   
#   There are two problems with this stat:
#   - The BaseSeq tracking logic in MediaPipeline doesn't account for multiple ssrcs
#     (rtx)
# ==============================================================================

diff -r 6b3195b0e545 -r 7888e4be3d7e dom/media/webrtc/jsapi/PeerConnectionImpl.cpp
--- a/dom/media/webrtc/jsapi/PeerConnectionImpl.cpp	Mon Sep 06 16:04:25 2021 +0200
+++ b/dom/media/webrtc/jsapi/PeerConnectionImpl.cpp	Wed Sep 08 10:34:57 2021 +0200
@@ -2609,11 +2609,13 @@
         // Gather pipeline stats.
         nsString localId = u"outbound_rtp_"_ns + idstr;
         nsString remoteId;
+
         Maybe<uint32_t> ssrc;
-        std::vector<unsigned int> ssrcvals =
-            aPipeline->mConduit->GetLocalSSRCs();
+        Maybe<uint16_t> base_seq;
+        std::vector<uint32_t> ssrcvals = aPipeline->mConduit->GetLocalSSRCs();
         if (!ssrcvals.empty()) {
           ssrc = Some(ssrcvals[0]);
+          base_seq = aPipeline->mConduit->RtpSendBaseSeqFor(ssrcvals[0]);
         }
 
         auto constructCommonRemoteInboundRtpStats =
@@ -2630,18 +2632,18 @@
                   kind);  // mediaType is the old name for kind.
               aRemote.mKind.Construct(kind);
               aRemote.mLocalId.Construct(localId);
-              aReportBlockData.apply([&](auto& aData) {
-                aPipeline->RtpSendBaseSeq().apply([&](uint32_t aBaseSeq) {
-                  if (aData.report_block().extended_highest_sequence_number <
-                      aBaseSeq) {
-                    aRemote.mPacketsReceived.Construct(0);
-                  } else {
-                    aRemote.mPacketsReceived.Construct(
-                        aData.report_block().extended_highest_sequence_number -
-                        aData.report_block().packets_lost - aBaseSeq + 1);
-                  }
-                });
-              });
+              if (aReportBlockData && base_seq) {
+                if (aReportBlockData->report_block()
+                        .extended_highest_sequence_number < *base_seq) {
+                  aRemote.mPacketsReceived.Construct(0);
+                } else {
+                  aRemote.mPacketsReceived.Construct(
+                      aReportBlockData->report_block()
+                          .extended_highest_sequence_number -
+                      aReportBlockData->report_block().packets_lost -
+                      *base_seq + 1);
+                }
+              }
             };
 
         auto constructCommonOutboundRtpStats =